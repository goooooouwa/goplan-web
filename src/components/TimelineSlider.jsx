import { Slider } from '@mui/material';
import React, { useEffect, useState } from 'react';

function valuetext(value) {
  return value;
}

export default function TimelineSlider(props) {
  const [value, setValue] = useState([props.rangeStart, props.rangeEnd]);

  const valueLabelFormat = (value) => {
    const index = props.marks.findIndex((mark) => mark.value === value);
    if (index === -1) {
      return;
    }
    return props.marks[index].label;
  };

  const handleChangeCommitted = (event, newValue) => {
    props.handleChangeCommited(value);
  };

  const handleChange = (event, newValue, activeThumb) => {
    if (!Array.isArray(newValue)) {
      return;
    }

    if (activeThumb === 0) {
      if (!props.disableRangeStart) {
        setValue(newValue);
      }
    } else {
      if (!props.disableRangeEnd) {
        setValue(newValue);
      }
    }
  };

  useEffect(()=>{
    setValue([props.rangeStart, props.rangeEnd]);
  },[props.rangeStart, props.rangeEnd]);

  return (
    <>
      <Slider
        valueLabelFormat={valueLabelFormat}
        value={value}
        disabled={props.disableRangeStart && props.disableRangeEnd}
        onChange={handleChange}
        onChangeCommitted={handleChangeCommitted}
        getAriaValueText={valuetext}
        valueLabelDisplay="auto"
        step={1}
        marks
        min={props.rangeMin}
        max={props.rangeMax}
      />
    </>
  );
}
