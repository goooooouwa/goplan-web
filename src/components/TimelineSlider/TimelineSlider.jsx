import { Slider } from '@mui/material';
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import SHARED_PROP_TYPES from 'utils/sharedPropTypes';

function valuetext(value) {
  return value;
}

export default function TimelineSlider(props) {
  const [value, setValue] = useState([props.rangeStart, props.rangeEnd]);

  const valueLabelFormat = (value) => {
    const index = props.marks.findIndex((mark) => mark.value === value);
    if (index === -1) {
      return;
    }
    return props.marks[index].label;
  };

  const handleChangeCommitted = (event, newValue) => {
    props.handleChangeCommited(filteredRangeValue(newValue));
  };

  const filteredRangeValue = (newValue) => {
    let [filteredRangeStart, filterRangeEnd] = value;
    if (!props.disableRangeStart) {
      filteredRangeStart = newValue[0];
    }
    if (!props.disableRangeEnd) {
      filterRangeEnd = newValue[1];
    }
    return [filteredRangeStart, filterRangeEnd];
  };

  const handleChange = (event, newValue, activeThumb) => {
    if (!Array.isArray(newValue)) {
      return;
    }

    setValue(filteredRangeValue(newValue));
  };

  useEffect(() => {
    setValue([props.rangeStart, props.rangeEnd]);
  }, [props.rangeStart, props.rangeEnd]);

  return (
    <>
      <Slider
        valueLabelFormat={valueLabelFormat}
        value={value}
        disabled={props.disableRangeStart && props.disableRangeEnd}
        onChange={handleChange}
        onChangeCommitted={handleChangeCommitted}
        getAriaValueText={valuetext}
        valueLabelDisplay="auto"
        step={1}
        marks
        min={props.rangeMin}
        max={props.rangeMax}
      />
    </>
  );
}

TimelineSlider.propTypes = {
  rangeMin: PropTypes.number.isRequired,
  rangeMax: PropTypes.number.isRequired,
  rangeStart: PropTypes.number.isRequired,
  rangeEnd: PropTypes.number.isRequired,
  disableRangeStart: PropTypes.bool,
  disableRangeEnd: PropTypes.bool,
  marks: SHARED_PROP_TYPES.marks,
  handleChangeCommited: PropTypes.func.isRequired
};