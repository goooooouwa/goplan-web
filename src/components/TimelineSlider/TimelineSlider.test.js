import React from 'react';
import renderer from 'react-test-renderer';
import TimelineSlider from './TimelineSlider';

it('changes the class when hovered', () => {
    const marks = [{
        value: 1,
        label: "Week 1"
    },{
        value: 2,
        label: "Week 2"
    }];
    const component = renderer.create(
        <TimelineSlider
            marks={marks}
            rangeMin={0}
            rangeMax={4}
            rangeStart={1}
            rangeEnd={3}
            disableRangeStart={true}
            disableRangeEnd={false}
            handleChangeCommited={() => {}}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // renderer.act(() => {
    //     tree.props.onMouseEnter();
    // });
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // renderer.act(() => {
    //     tree.props.onMouseLeave();
    // });
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
});