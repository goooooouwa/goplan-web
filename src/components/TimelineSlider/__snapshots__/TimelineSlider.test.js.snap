// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `
<span
  className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
  onMouseDown={[Function]}
>
  <span
    className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
  />
  <span
    className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
    style={
      Object {
        "left": "25%",
        "width": "50%",
      }
    }
  />
  <span
    className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
    data-index={0}
    style={
      Object {
        "left": "0%",
      }
    }
  />
  <span
    className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
    data-index={1}
    style={
      Object {
        "left": "25%",
      }
    }
  />
  <span
    className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
    data-index={2}
    style={
      Object {
        "left": "50%",
      }
    }
  />
  <span
    className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
    data-index={3}
    style={
      Object {
        "left": "75%",
      }
    }
  />
  <span
    className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
    data-index={4}
    style={
      Object {
        "left": "100%",
      }
    }
  />
  <span
    className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
    data-index={0}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    style={
      Object {
        "left": "25%",
        "pointerEvents": undefined,
      }
    }
  >
    <input
      aria-orientation="horizontal"
      aria-valuemax={4}
      aria-valuemin={0}
      aria-valuenow={1}
      aria-valuetext={1}
      data-index={0}
      disabled={false}
      max={4}
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      step={1}
      style={
        Object {
          "border": 0,
          "clip": "rect(0 0 0 0)",
          "direction": "ltr",
          "height": "100%",
          "margin": -1,
          "overflow": "hidden",
          "padding": 0,
          "position": "absolute",
          "whiteSpace": "nowrap",
          "width": "100%",
        }
      }
      type="range"
      value={1}
    />
    <span
      aria-hidden={true}
      className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
    >
      <span
        className="MuiSlider-valueLabelCircle"
      >
        <span
          className="MuiSlider-valueLabelLabel"
        >
          Week 1
        </span>
      </span>
    </span>
  </span>
  <span
    className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
    data-index={1}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    style={
      Object {
        "left": "75%",
        "pointerEvents": undefined,
      }
    }
  >
    <input
      aria-orientation="horizontal"
      aria-valuemax={4}
      aria-valuemin={0}
      aria-valuenow={3}
      aria-valuetext={3}
      data-index={1}
      disabled={false}
      max={4}
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      step={1}
      style={
        Object {
          "border": 0,
          "clip": "rect(0 0 0 0)",
          "direction": "ltr",
          "height": "100%",
          "margin": -1,
          "overflow": "hidden",
          "padding": 0,
          "position": "absolute",
          "whiteSpace": "nowrap",
          "width": "100%",
        }
      }
      type="range"
      value={3}
    />
    <span
      aria-hidden={true}
      className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
    >
      <span
        className="MuiSlider-valueLabelCircle"
      >
        <span
          className="MuiSlider-valueLabelLabel"
        />
      </span>
    </span>
  </span>
</span>
`;
